Creating tcp handler without socket should not be permitted

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
<!--- If you're looking for help, please see https://sensuapp.org/support for resources --->
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
Since a tcp handler requires a socket attribute, I would expect `sensuctl create` to fail in some way, or for `sensuctl handler list` to report invalid configurations.

## Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
I can create a handler that crashes `sensuctl`

## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas as to the implementation of the addition or change -->
Add some error checking in sensuctl? (That's kind of naive suggestion not knowing this codebase well)

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code or configuration to reproduce, if relevant -->
1. Create a json object file with the following spec, lacking socket attribute:

handler.json:
```json
{
  "type": "handler",
  "spec": {
    "name": "tcp_handler",
    "environment": "default",
    "organization": "default",
    "timeout": 30,
    "type": "tcp"
  }
}
```
2. Create the object
```
# sensuctl create -f handler.json
```
3. Try to list it
```
# sensuctl handler list
```
4. Observe the panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x88f324]

goroutine 1 [running]:
github.com/sensu/sensu-go/cli/commands/handler.printToTable.func6(0x9a0760, 0xc420238000, 0x4, 0x4)
        /go/src/github.com/sensu/sensu-go/cli/commands/handler/list.go:102 +0x234
github.com/sensu/sensu-go/cli/elements/table.(*Table).writeRow(0xc42004fbd8, 0xc42021e040)
        /go/src/github.com/sensu/sensu-go/cli/elements/table/standard.go:90 +0xcb
github.com/sensu/sensu-go/cli/elements/table.(*Table).writeRows(0xc42004fbd8, 0x8f8da0, 0xc4201f4160)
        /go/src/github.com/sensu/sensu-go/cli/elements/table/standard.go:83 +0x24e
github.com/sensu/sensu-go/cli/elements/table.(*Table).Render(0xc42004fbd8, 0xa31400, 0xc420096008, 0x8f8da0, 0xc4201f4160)
        /go/src/github.com/sensu/sensu-go/cli/elements/table/standard.go:66 +0x82
github.com/sensu/sensu-go/cli/commands/handler.printToTable(0x8f8da0, 0xc4201f4160, 0xa31400, 0xc420096008)
        /go/src/github.com/sensu/sensu-go/cli/commands/handler/list.go:132 +0x30a
github.com/sensu/sensu-go/cli/commands/helpers.Print(0xc4201a3400, 0xc4200930f0, 0x7, 0x9f21d0, 0x8f8da0, 0xc4201f4160, 0x0, 0x0)
        /go/src/github.com/sensu/sensu-go/cli/commands/helpers/print.go:23 +0xb3
github.com/sensu/sensu-go/cli/commands/handler.ListCommand.func1(0xc4201a3400, 0xd14408, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/sensu/sensu-go/cli/commands/handler/list.go:41 +0x252
github.com/sensu/sensu-go/vendor/github.com/spf13/cobra.(*Command).execute(0xc4201a3400, 0xd14408, 0x0, 0x0, 0xc4201a3400, 0xd14408)
        /go/src/github.com/sensu/sensu-go/vendor/github.com/spf13/cobra/command.go:753 +0x468
github.com/sensu/sensu-go/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420180280, 0xc4201b0f00, 0xc4201b2000, 0xc4201b2280)
        /go/src/github.com/sensu/sensu-go/vendor/github.com/spf13/cobra/command.go:843 +0x30a
github.com/sensu/sensu-go/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420180280, 0xc4201473e0, 0xc4200d9628)
        /go/src/github.com/sensu/sensu-go/vendor/github.com/spf13/cobra/command.go:791 +0x2b
main.main()
        /go/src/github.com/sensu/sensu-go/cli/cmd/start.go:26 +0xbd
```
## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context (e.g. links to configuration settings, stack strace or log data) helps us come up with a solution that is most useful in the real world -->
How has this issue affected you? Minor pain, blurred vision, slight headache. 
What are you trying to accomplish? Writing custom chef resources to manage this stuff.

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Sensu version used (sensuctl, sensu-backend, and/or sensu-agent): `sensu-ctl version dev-nightly#9bba4cf, build 9bba4cf6af864bbabf0a3120ec2ad6d187552a8b, built 2018-05-06T09:24:10+0000`
* Installation method (packages, binaries, docker etc.): nightly packages from packagecloud
* Operating System and version (e.g. Ubuntu 14.04): Ubuntu 16.04

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,7 +14,7 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 - Add logging around the Sensu event pipeline
 
 ### Fixed
-
+- Fix handler validation routine
 
 ## [2.0.0-beta.1] - 2018-05-07
 ### Added

--- a/types/handler.go
+++ b/types/handler.go
@@ -31,8 +31,8 @@ func (h *Handler) Validate() error {
 		return errors.New("handler name " + err.Error())
 	}
 
-	if err := validateHandlerType(h.Type); err != nil {
-		return errors.New("handler type " + err.Error())
+	if err := h.validateType(); err != nil {
+		return err
 	}
 
 	if h.Environment == "" {
@@ -43,6 +43,34 @@ func (h *Handler) Validate() error {
 		return errors.New("organization must be set")
 	}
 
+	return nil
+}
+
+func (h *Handler) validateType() error {
+	if h.Type == "" {
+		return errors.New("empty handler type")
+	}
+
+	switch h.Type {
+	case "pipe", "set", "grpc":
+		return nil
+	case "tcp", "udp":
+		return h.Socket.Validate()
+	}
+
+	return fmt.Errorf("unknown handler type: %s", h.Type)
+}
+
+func (s *HandlerSocket) Validate() error {
+	if s == nil {
+		return errors.New("tcp and udp handlers need a valid socket")
+	}
+	if len(s.Host) == 0 {
+		return errors.New("socket host undefined")
+	}
+	if s.Port == 0 {
+		return errors.New("socket port undefined")
+	}
 	return nil
 }
 

--- a/types/handler_test.go
+++ b/types/handler_test.go
@@ -1,9 +1,11 @@
 package types
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestFixtureHandler(t *testing.T) {
@@ -28,24 +30,170 @@ func TestFixtureSocketHandler(t *testing.T) {
 }
 
 func TestHandlerValidate(t *testing.T) {
-	var h Handler
+	tests := []struct {
+		Handler Handler
+		Error   string
+	}{
+		{
+			Handler: Handler{},
+			Error:   "handler name must not be empty",
+		},
+		{
+			Handler: Handler{
+				Name: "foo",
+			},
+			Error: "empty handler type",
+		},
+		{
+			Handler: Handler{
+				Name: "foo",
+				Type: "pipe",
+			},
+			Error: "environment must be set",
+		},
+		{
+			Handler: Handler{
+				Name:        "foo",
+				Type:        "pipe",
+				Environment: "default",
+			},
+			Error: "organization must be set",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "pipe",
+				Organization: "default",
+				Environment:  "default",
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "grpc",
+				Organization: "default",
+				Environment:  "default",
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "grpc",
+				Organization: "default",
+				Environment:  "default",
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "set",
+				Organization: "default",
+				Environment:  "default",
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "tcp",
+				Organization: "default",
+				Environment:  "default",
+			},
+			Error: "tcp and udp handlers need a valid socket",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "tcp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Host: "localhost",
+				},
+			},
+			Error: "socket port undefined",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "tcp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Port: 1234,
+				},
+			},
+			Error: "socket host undefined",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "tcp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Host: "localhost",
+					Port: 1234,
+				},
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "udp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Host: "localhost",
+				},
+			},
+			Error: "socket port undefined",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "udp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Port: 1234,
+				},
+			},
+			Error: "socket host undefined",
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "udp",
+				Organization: "default",
+				Environment:  "default",
+				Socket: &HandlerSocket{
+					Host: "localhost",
+					Port: 1234,
+				},
+			},
+		},
+		{
+			Handler: Handler{
+				Name:         "foo",
+				Type:         "magic",
+				Organization: "default",
+				Environment:  "default",
+			},
+			Error: "unknown handler type: magic",
+		},
+	}
 
-	// Invalid name
-	assert.Error(t, h.Validate())
-	h.Name = "foo"
-
-	// Invalid type
-	assert.Error(t, h.Validate())
-	h.Type = "pipe"
-
-	// Invalid organization
-	assert.Error(t, h.Validate())
-	h.Organization = "default"
-
-	// Invalid environment
-	assert.Error(t, h.Validate())
-	h.Environment = "default"
-
-	// Valid handler
-	assert.NoError(t, h.Validate())
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%v", test.Handler), func(t *testing.T) {
+			if err := test.Handler.Validate(); err != nil {
+				if len(test.Error) > 0 {
+					require.Equal(t, test.Error, err.Error())
+				} else {
+					t.Fatal(err)
+				}
+			} else if len(test.Error) > 0 {
+				t.Fatal("expected error, got none")
+			}
+		})
+	}
 }

--- a/types/validators.go
+++ b/types/validators.go
@@ -22,24 +22,6 @@ var StrictNameRegex = regexp.MustCompile(`\A[a-z0-9\_\.\-]+\z`)
 // entity:foo)
 var SubscriptionNameRegex = regexp.MustCompile(`\A[\w\.\-]+\:?[\w\.\-]+\z`)
 
-func validateHandlerType(t string) error {
-	if t == "" {
-		return errors.New("must not be empty")
-	}
-
-	switch t {
-	case
-		"pipe",
-		"tcp",
-		"udp",
-		"transport",
-		"set":
-		return nil
-	}
-
-	return errors.New("is unknown")
-}
-
 // ValidateName validates the name of an element so it's not empty and it does
 // not contains specical characters. Compatible with Sensu 1.0.
 func ValidateName(name string) error {

--- a/types/validators_test.go
+++ b/types/validators_test.go
@@ -6,16 +6,6 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestValidateHandlerType(t *testing.T) {
-	assert.Error(t, validateHandlerType(""))
-	assert.Error(t, validateHandlerType("foo"))
-	assert.NoError(t, validateHandlerType("pipe"))
-	assert.NoError(t, validateHandlerType("tcp"))
-	assert.NoError(t, validateHandlerType("udp"))
-	assert.NoError(t, validateHandlerType("transport"))
-	assert.NoError(t, validateHandlerType("set"))
-}
-
 func TestValidateName(t *testing.T) {
 	assert.Error(t, ValidateName(""))
 	assert.Error(t, ValidateName("foo bar"))

